File: ocaml_aoc/lib/2023/day5.ml
open Core

let parse_seeds_normal seeds =
  seeds
  |> Str.replace_first (Str.regexp "seeds: ") ""
  |> String.split ~on:' '
  |> List.map ~f:Int.of_string
;;

let rec generate_seeds_range (seeds_list : string list) =
  match seeds_list with
  | [] -> []
  | [ _ ] -> failwith "Invalid seeds"
  | start :: length :: rest -> [ start; length ] :: generate_seeds_range rest
;;

let parse_seeds_range seeds =
  seeds
  |> Str.replace_first (Str.regexp "seeds: ") ""
  |> String.split ~on:' '
  |> generate_seeds_range
  |> List.map ~f:(List.map ~f:Int.of_string)
;;

let parse_data input ~seed_parser =
  match input |> Str.split (Str.regexp "\n\n") with
  | [] -> failwith "No data"
  | seeds :: maps ->
    ( seed_parser seeds
    , maps
      |> List.map ~f:(String.split ~on:'\n')
      |> List.map ~f:List.tl_exn
      |> List.map ~f:(List.map ~f:(String.split ~on:' '))
      |> List.map ~f:(List.map ~f:(List.map ~f:Int.of_string)) )
;;

let rec find_location loc (maps : int list list list) =
  match maps with
  | [] -> loc
  | each_map :: rest_of_maps ->
    (match each_map with
     | [] -> find_location loc rest_of_maps
     | each_row :: rest_of_rows ->
       (match each_row with
        | [] -> failwith "Invalid map"
        | [ dest; src; length ] ->
          let max_src = src + length in
          if loc >= src && loc <= max_src
          then (
            let new_loc = loc + (dest - src) in
            find_location new_loc rest_of_maps)
          else find_location loc (rest_of_rows :: rest_of_maps)
        | _ -> failwith "Invalid map"))
;;

let find_lowest_location ((seeds : int list), (maps : int list list list)) =
  List.fold
    ~init:0
    ~f:(fun lowest seed ->
      match lowest with
      | 0 -> find_location seed maps
      | x ->
        let loc = find_location seed maps in
        if loc < x then loc else x)
    seeds
;;

let part_one input =
  let seeds, maps = parse_data input ~seed_parser:parse_seeds_normal in
  find_lowest_location (seeds, maps)
;;

let rec find_lowest_location_v2 loc ((seeds : int list list), (maps : int list list list))
  =
  match seeds with
  | [] -> loc
  | seed_range :: rest ->
    let seed_start = List.hd_exn seed_range in
    let seed_length = List.nth_exn seed_range 1 in
    let range = List.init (seed_length - 1) ~f:(fun i -> seed_start + i) in
    printf
      "Range, %d-%d, start: %d length: %d\n"
      (List.hd_exn range)
      (List.last_exn range)
      seed_start
      seed_length;
    let new_loc = find_lowest_location (range, maps) in
    if loc = 0 || new_loc < loc
    then find_lowest_location_v2 new_loc (rest, maps)
    else find_lowest_location_v2 loc (rest, maps)
;;

let part_two input =
  let seeds, maps = parse_data input ~seed_parser:parse_seeds_range in
  find_lowest_location_v2 0 (seeds, maps)
;;

(* find_lowest_location (seeds, maps) *)

module Exec = struct
  let run () =
    print_endline "Day 5:";
    let input = Helper.FileReader.read_file "../inputs/2023/day5-puzzle.txt" in
    printf "Part 1: %d\n" (part_one input);
    printf "Part 2: %d\n" (part_two input);
    print_endline "-------------------------"
  ;;
end

File: go-aoc/2023/day5.go
package aoc_2023

import (
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"

	"github.com/andrasat/aoc/utils"
)

type Map struct {
	Dest int
	Src  int
	Size int
}

type Data struct {
	seeds []int
	maps  [][]Map
}

func parseData(input string, getSeedRange bool) Data {
	data := Data{
		seeds: []int{},
		maps:  [][]Map{},
	}

	sections := strings.Split(input, "\n\n")

	seeds, maps := sections[0], sections[1:]

	if getSeedRange {
		separated_seeds := strings.Split(strings.Replace(seeds, "seeds: ", "", 1), " ")
		for i, seed := range separated_seeds {
			intSeed, _ := strconv.Atoi(seed)

			if (i+1)%2 == 0 {
				start := data.seeds[i-1] + 1
				for j := start; j < start+intSeed; j++ {
					if exist := slices.ContainsFunc(data.seeds, func(seed int) bool {
						return seed == j
					}); !exist {
						data.seeds = append(data.seeds, j)
					}
				}
			} else {
				data.seeds = append(data.seeds, intSeed)
			}
		}
	} else {
		separated_seeds := strings.Split(strings.Replace(seeds, "seeds: ", "", 1), " ")
		for _, seed := range separated_seeds {
			intSeed, _ := strconv.Atoi(seed)
			data.seeds = append(data.seeds, intSeed)
		}
	}

	regexPattern := regexp.MustCompile(`\d+`)

	filteredMaps := [][]Map{}
	for i, mapSection := range maps {
		rows := strings.Split(mapSection, "\n")[1:]

		filteredMaps = append(filteredMaps, []Map{})

		for _, row := range rows {
			if ok := regexPattern.MatchString(row); ok {
				map_data := strings.Split(row, " ")
				dest, _ := strconv.Atoi(map_data[0])
				src, _ := strconv.Atoi(map_data[1])
				size, _ := strconv.Atoi(map_data[2])
				filteredMaps[i] = append(filteredMaps[i], Map{Dest: dest, Src: src, Size: size})
			}
		}
	}

	data.maps = filteredMaps

	return data
}

func findLowestLocation(data Data) int {
	locations := []int{}

	for _, seed := range data.seeds {
		currentLocation := seed

		for _, mapSection := range data.maps {
			for _, m := range mapSection {

				maxSrc := m.Src + m.Size

				if currentLocation >= m.Src && currentLocation <= maxSrc {
					newLoc := currentLocation + (m.Dest - m.Src)
					currentLocation = newLoc
					break
				}
			}
		}

		locations = append(locations, currentLocation)
	}

	return slices.Min(locations)
}

func runDay5() error {
	fmt.Println("\n ----- DAY 5 -----")

	input, err := utils.Readfile("./../inputs/2023/day5-puzzle.txt")
	if err != nil {
		return err
	}

	fmt.Printf("Part: 1 - %d\n", findLowestLocation(parseData(input, false)))
	fmt.Printf("Part: 2 - %d\n", findLowestLocation(parseData(input, true)))

	return nil
}
